pydantic/utils.py

```diff
@@ -525,38 +465,114 @@ def _normalize_indexes(
+                raise TypeError(f'Unexpected type of exclude value for index "{i}" {v.__class__}')
+            if i == '__all__':
+                all_items = self._coerce_value(v)
+                continue
+            if not isinstance(i, int):
+                raise TypeError(
+                    'Excluding fields from a sequence of sub-models or dicts must be performed index-wise: '
+                    'expected integer keys or keyword "__all__"'
+                )
+            normalized_i = v_length + i if i < 0 else i
+            normalized_items[normalized_i] = self.merge(v, normalized_items.get(normalized_i))
+
+        if not all_items:
+            return normalized_items
+        if all_items is ...:
+            for i in range(v_length):
+                normalized_items.setdefault(i, ...)
             return normalized_items
+        for i in range(v_length):
+            normalized_item = normalized_items.setdefault(i, {})
+            if normalized_item is not ...:
+                normalized_items[i] = self.merge(all_items, normalized_item)
+        return normalized_items
+
+    @overload
+    @classmethod
+    def merge(
+        cls, base: Union['AbstractSetIntStr', 'MappingIntStrAny'], override: Optional[Any], intersect: bool = False
+    ) -> 'MappingIntStrAny':
+        ...
+
+    @overload
+    @classmethod
+    def merge(
+        cls, base: Optional[Any], override: Union['AbstractSetIntStr', 'MappingIntStrAny'], intersect: bool = False
+    ) -> 'MappingIntStrAny':
+        ...
+
+    @overload
+    @classmethod
+    def merge(cls, base: None, override: None, intersect: bool = False) -> None:
+        ...
+
+    @overload
+    @classmethod
+    def merge(cls, base: Optional[Any], override: Optional[Any], intersect: bool = False) -> Optional[Any]:
+        ...
+
+    @classmethod
+    def merge(cls, base: Any, override: Any, intersect: bool = False) -> Any:
+        """
+        Merge a ``base`` item with an ``override`` item.
+
+        Both ``base`` and ``override`` are converted to dictionaries if possible.
+        Sets are converted to dictionaries with the sets entries as keys and
+        Ellipsis as values.
+
+        Each key-value pair existing in ``base`` is merged with ``override``,
+        while the rest of the key-value pairs are updated recursively with this function.
+
+        Merging takes place based on the "union" of keys if ``intersect`` is
+        set to ``False`` (default) and on the intersection of keys if
+        ``intersect`` is set to ``True``.
+        """
+        override = cls._coerce_value(override)
+        base = cls._coerce_value(base)
+        if override is None:
+            return base
+        if base is ... or base is None:
+            return override
+        elif override is ...:
+            return base if intersect else override
+
+        if intersect:
+            merge_keys = override.keys() & base.keys()
```

Comment:
It looks like retaining order can't be guaranteed. I can't work out if that could be a problem? Might be worth adding a note to the docstring?

Reply:
Yeah it is if we're using numeric indexes (for example keys -1 and 1).  
Don't know if the slight performance benefit is worth inconsistent behavior in some edge cases. 

I changed the code to preserve ordering for now (if you'd rather keep the direct set operations I can change it back).